libraries/ms-common/src/stm32f0xx/interrupt.c

```diff
@@ -0,0 +1,200 @@
+#include "interrupt.h"
+
+#include <stdint.h>
+#include <stdlib.h>
+
+#include "gpio.h"
+#include "gpio_cfg.h"
+#include "status.h"
+#include "stm32f0xx_exti.h"
+#include "stm32f0xx_misc.h"
+#include "stm32f0xx_rcc.h"
+#include "stm32f0xx_syscfg.h"
+
+#define NUM_INTERRUPT_GPIO_CHANNELS 3
+static InterruptPriority s_interrupt_gpio_priority_map[NUM_INTERRUPT_GPIO_CHANNELS];
+
+void interrupt_init() {
+  RCC_APB2PeriphClockCmd(RCC_APB2Periph_SYSCFG, ENABLE);
+
+  for (int i = 0; i < NUM_INTERRUPT_GPIO_CHANNELS; i++) {
+    s_interrupt_gpio_priority_map[i] = NUM_INTERRUPT_PRIORITY;
+  }
+}
+
+#pragma GCC push_options
+#pragma GCC optimize("O0")
+// Use this to count how many nested disables there are and only enable once out of all critical
+// sections.
+static uint8_t s_interrupt_priority_mask = 0;
+
+void interrupt_enable() {
+  if (s_interrupt_priority_mask > 0) {
```

Comment:
Fair, I don't really want people to have to worry about whether or not interrupts are enabled. Hence why I went with this approach. It seems robust to me so long as people remember to call both sides of it... ? I'm not sure I follow. Same priority interrupts get queued while higher priority interrupts would interrupt the ISR anyway. Enabling them shouldn't change that.